File: sections/05_sorting_as_tool/problems/b_rank_table/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Row {
    string name;
    int score;
    int penalty;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        cin >> n;
        vector<Row> rows(n);
        for (auto &row : rows) {
            cin >> row.name >> row.score >> row.penalty;
        }
        sort(rows.begin(), rows.end(), [](const Row &a, const Row &b) {
            if (a.score != b.score) return a.score > b.score;
            if (a.penalty != b.penalty) return a.penalty < b.penalty;
            return a.name < b.name;
        });
        for (int i = 0; i < n; ++i) {
            if (i) cout << ' ';
            cout << rows[i].name;
        }
        cout << '\n';
    }

    return 0;
}

File: sections/05_sorting_as_tool/problems/b_rank_table/solution.py
"""Reference solution for B. Rank Table."""

import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    t = int(data[0])
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = int(data[idx])
        idx += 1
        rows: list[tuple[str, int, int]] = []
        for _ in range(n):
            name = data[idx].decode()
            score = int(data[idx + 1])
            penalty = int(data[idx + 2])
            idx += 3
            rows.append((name, score, penalty))
        rows.sort(key=lambda row: (-row[1], row[2], row[0]))
        out.append(" ".join(row[0] for row in rows))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

